### ISSUE TITLE
Discussion on strategy to parallelize deployments focused on GitHub workflows/jobs

### ISSUE THREAD

**Issue by consideRatio** [2i2c-org/infrastructure#1131]
### Summary

We are trying to decide how we should structure our GitHub Actions workflow to deploy hubs in parallel. This will determine what inputs into the `strategy.matrix` fields will be required and thus will unblock #1112 

- Should the support chart and `cert-manager` be deployed sequentially with the hubs or can it be done in parallel? What are the implications of either scenario?
- We want failed deploys to staging to prevent deploys to prod hubs (canary deploys). Should a failure for staging on _one_ cluster stop prod deploys to _all other_ clusters? If no, what's the simplest way of avoiding this pattern?

---

These are the strategies we consider in GitHub's CI system on how to deploy multiple hubs in parallel etc.

These strategies influences the logic we may want to embed in the deployer script. Due to that, I think now is the time to clarify a strategy to try use in the GitHub CI context.

## Current situation

We have [a GitHub workflow](https://github.com/2i2c-org/infrastructure/blob/master/.github/workflows/deploy-hubs.yaml) with a single job definition running multiple times (`strategy.matrix`), once per cluster. For each cluster, it unconditionally re-deploys everything in sequence.

## Discussion goals

- Clarify strategies we consider
- Reduce the amount of strategies we consider
- Decide to go for a specific strategy

If we manage to reach these goals I think development/review of #1112 will be easier.

## Alternative strategies

In the comments below are strategies I've tried to make explicit to some degree, with a focus on what is done in GitHub's workflow files as that influences what we need from the deployer script.

A key strategy to be aware about is that we can dispatch separate workflows to run in order to run something in parallel. The key downside of doing this is that GitHub's UI don't know we have dispatched a workflow and it will show up decoupled from the workflow that dispatched it to run. To see what I mean with this practically, you can inspect [this workflow execution](https://github.com/consideRatio/testing-gh-workflows/runs/5584159429?check_suite_focus=true) that finished after dispatching more workflows as can be seen [in the list showing all workflows](https://github.com/consideRatio/testing-gh-workflows/actions).

**Comment by consideRatio** [comment]
## Single workflow: pre-planner -> cluster setup -> parallel hub deployment

This strategy would have a single GitHub workflow with a few jobs to execute in sequence, most of them running with multiple variations using `strategy.matrix`.

**Job 1 - The pre-planner job**
This job would calculate what is put in `strategy.matrix` on the remaining two jobs, which is deciding what clusters to re-deploy something in, and what to re-deploy in those clusters.

**Job 2 - Cluster setup**
This would would run for each cluster the planner job concluded it needs to run for. Where it conditionally would setup the cluster central parts such as optionally re-deploying a `support`/`staging`/`dask-staging` chart.

**Job 3a - The re-planner job (not verified to be possible)** 
This job could, if possible, filter out hubs that shouldn't deploy because of a cluster setup failure in Job 2, allowing Job 3b to deploy some of the hubs instead of halting everything because of any cluster setup failure.

Work is needed to verify this would be possible at all, because it depends on the ability to distinguish failures of specific `strategy.matrix` variations of Job 2 which I'm not sure is possible as `needs.<job-id>.result` is just a single string for the entire job definition. I have some hope it would be possible using job outputs but I'm not sure.

Related:
- https://docs.github.com/en/actions/using-jobs/defining-outputs-for-jobs
- https://docs.github.com/en/actions/learn-github-actions/contexts#needs-context

**Job 3b - Parallel hub deployment**
This job would run once for each hub in any of the clusters that needed to be re-deployed.

### Pros and cons

No GitHub UI drawback of decoupling the initial workflow from what ends up running
All clusters hub deployments is delayed waiting for other clusters setup
Would need to abort all parallel hub deployments unless job 3a works out whenever another cluster setup would fail

**Comment by consideRatio** [comment]
### Initial workflow that quickly dispatches cluster scoped workflows

This strategy would let an initial workflow quickly decide if a cluster needs any re-deployment operation, and if so, dispatch a dedicated workflow for it.

The dedicated workflow could thereafter have logic to decide what it should do scoped specifically at a single cluster.

#### Pros and cons

- Not stuck waiting for other clusters
- No multi cluster logic needed (See job 3a)
- GitHub UI drawback of decoupling the initial workflow from what ends up running

and then make a decision if it needs to dispatch new workflows to run in parallel, once per cluster it deems has changed in any way. Then, the dispatched workflow can be declared with clearly defined inputs etc that clarifies what should be done, and could be run only for a specific cluster.

**Comment by consideRatio** [comment]
### Abdonned strategy: cluster setup -> workflow dispatch for parallel hub deployment

Actually I think this doesn't need to be discussed any more as it was planned based on a false assumption that we needed to be able to handle multiple staging/prod combinations of hubs in the same cluster. It turns out though, all our staging hubs staging hubs for the clusters, not for individual hubs in the cluster.

Briefly though, in this I just had a single job that run once per cluster, and dispatched new workflows to run `curl` and a private access token. See https://github.com/consideRatio/testing-gh-workflows/blob/main/.github/workflows/chain_dispatcher.yaml for details.

**Comment by sgibson91** [comment]
I think this is my vision

![CICD-map](https://user-images.githubusercontent.com/44771837/159114445-bbd260d7-79ec-4cfa-8784-b553b75e1ae1.png)

**Comment by consideRatio** [comment]
@sgibson91 thank you for putting in the effort to communicate this clearly!

I'm not sure I understand the details regarding:
- involved workflow file(s)
- workflows' jobs' `needs` and `strategy.matrix` configurations

Complexity:
- _I think this description _must_ be a multi-workflow._
  I think this because the "for each cluster" section included two separate jobs with an arrow to another job. You can't in a single workflow have dynamically many "separate job chains" which I perceive this to declare unless "for each cluster" implies a dedicated workflow that is dispatched.
  Note that you can't have `strategy.matrix` when letting a job reference a _reusable workflow_, if you could, this would have been possible and very nice all within a single workflow.
- _My understanding from this is that you would have the (cert-manager + support chart) upgrade run in parallel to the staging hub upgrades -> prod hub upgrades._
  Isn't this problematic to do in parallel, and something that we should do in sequence instead?

**Comment by sgibson91** [comment]
I'm going to put up my original idea of how this workflow would look. I don't think it's perfect, but I'm at least comfortable it will work 😂 I envision this to run in a single GitHub Actions workflow file.

### Job 1: Plan jobs

Based on changed files, decide which clusters need something redeploying in, and what exactly needs redeploying. Output some environment variables that describe these jobs when read by the `strategy.matrix` field.

### Job 2: Upgrade support chart

```
needs: [job_1]
strategy.matrix = [{cluster_name: some_cluster}, ...]
```

Upgrade the support chart for each cluster that requires it, in parallel.

#### Note

> My understanding from this is that you would have the (cert-manager + support chart) upgrade run in parallel to the staging hub upgrades -> prod hub upgrades. Isn't this problematic to do in parallel, and something that we should do in sequence instead?

I don't know. My assumption was that if we thought it would be safe to simultaneously deploy to different prod hub namespaces, then the same would be true for the `cert-manager`/`support` namespaces. It might be problematic when we very first setup a cluster, but we tend to install the charts manually then? What do we think could go wrong here?

### Job 3: Upgrade staging hub(s)

```
needs: [job_1]
strategy.matrix = [{cluster_name: some_cluster}, ...]
```

Upgrade the staging hub deployed on a given cluster in parallel. We only upgrade the staging hubs on clusters that will also have production hubs upgraded, this will be decided in Job 1. (Special case for 2i2c which has staging and dask-staging.)

### Job 4: Upgrade production hubs

```
needs: [job_1, job_3]
strategy.matrix = [{cluster_name: some_cluster, hub_name: hub_1}, {cluster_name: some_cluster, hub_name: hub_2}, {cluster_name: different_cluster, hub_name: hub_3}, ...]
```

Upgrade the production hubs that require an upgrade for each cluster in parallel.

### Advantages

Having the staging upgrade be a separate job like this means we could use a [test-this-pr workflow](https://blog.jupyter.org/how-i-automated-authorised-cloud-deployments-from-pull-requests-with-github-actions-13f890538e32) to test PRs on staging hubs before merging

### Disadvantages

A failure in job 3 (staging hubs) would prevent any jobs in job 4 (prod hubs) from executing

**Comment by consideRatio** [comment]
@sgibson91 ah I think this mapped quite well to https://github.com/2i2c-org/infrastructure/issues/1131#issuecomment-1072573736 with some detail changes!

## About deploying support in parallel to staging

> > My understanding from this is that you would have the (cert-manager + support chart) upgrade run in parallel to the staging hub upgrades -> prod hub upgrades. Isn't this problematic to do in parallel, and something that we should do in sequence instead?
> 
> I don't know. My assumption was that if we thought it would be safe to simultaneously deploy to different prod hub namespaces, then the same would be true for the `cert-manager`/`support` namespaces. It might be problematic when we very first setup a cluster, but we tend to install the charts manually then? What do we think could go wrong here?

First of all, if we could do these in parallel, we would save a lot of time I think, and that would be great!

What does the support chart include, and would our staging/prod charts a) deploy and b) handle health tests without the support chart installed and/or upgraded to its latest version?

### Misc good to have helm charts

```yaml
description: Cluster wide depdencies for deployed hubs

dependencies:
  # Prometheus for collection of metrics.
  # https://github.com/prometheus-community/helm-charts/tree/main/charts/prometheus
  - name: prometheus
    version: 13.8.0
    repository: https://prometheus-community.github.io/helm-charts

  # Grafana for dashboarding of metrics.
  # https://github.com/grafana/helm-charts/tree/main/charts/grafana
  - name: grafana
    version: 6.11.0
    repository: https://grafana.github.io/helm-charts

  # ingress-nginx for a k8s Ingress resource controller that routes traffic from
  # a single IP entrypoint to various services exposed via k8s Ingress resources
  # that references this controller.
  - name: ingress-nginx
    version: 3.33.0
    repository: https://kubernetes.github.io/ingress-nginx

  # cluster-autoscaler for k8s clusters where it doesn't come out of the box (EKS)
  # https://github.com/kubernetes/autoscaler/tree/master/charts/cluster-autoscaler
  - name: cluster-autoscaler
    version: 9.10.3
    repository: https://kubernetes.github.io/autoscaler
    condition: cluster-autoscaler.enabled

  # In-cluster NFS server provisioner
  - name: nfs-server-provisioner
    version: 1.3.1
    repository: https://kvaps.github.io/charts
    condition: nfs-server-provisioner.enabled
```

### Misc k8s resources in dedicated Helm chart templates

![image](https://user-images.githubusercontent.com/3837114/159340268-b21fa840-dba7-4480-b527-dc1f0aad3a20.png)

### Reflections - is it acceptable to do support + staging at the same time?

I think it could be worth testing as a time saver as the time saved to deploy may be reduced significantly, but it is a bit of a risk as the helm charts staging/prod may depend on services provided by the support charts - even though they are independent helm charts.

I think nginx-ingress/cert-manager will be required for networking to reach staging. If they are not installed at all yet that is a problem for the health check, and if they are upgraded in parallel we may observe issues such as nginx-ingress controller breaking a connection or not being available while we run health checks.

I doubt there will be issues with cluster-autoscaler, prometheus, grafana, or even nfs-server-provisioner with regards to race conditions or similar but nfs-server-provisioner.

Ah okay well since it is support + staging at the same time, and not support + prod, and prod will wait for both support + staging, i think this is a good idea. The worst thing that happens is that we messed up a staging hub health check or similar - but not even the prods.

We trade complexity for speed, but the speed will be significant at least when we don't bottleneck on the amount of parallel runners we can get from GitHub. If we do that, it would probably have been better to do it in sequence. We will very very likely start more jobs than we can run simultaneously if we re redploy all charts, but I think we should be very restrictive about when we do that - leaning even towards explicitly declaring that instead of doing it by default if for example the deployer script changes.

**Comment by consideRatio** [comment]
## Standalone consideration

It would be great if you could re-run failed health checks without re-running a deployment operation, and if we could distinguish from the GitHub UI if the deployment succeeded vs failed and if the health check succeeded or failed.

It would be quite easy to do this by having a dedicated _step_ within a _job_, but it would be quite hard or come with notable compromises to do this with a dedicated _job_ within a _workflow_. I think... Hmmm...

Well actually... We could have a last job that would run prod healthchecks that would always run no matter what if the previous job succeeded. Hmm...

**Comment by yuvipanda** [comment]
Thank you so much for all the work you've put into this, @sgibson91 and @consideRatio!

@sgibson91 your comment in https://github.com/2i2c-org/infrastructure/issues/1131#issuecomment-1074232509 sounds very good to me as a tangible next step to do! The only change I'd like to see is to deploy the support chart (only if it changes?) *not* in parallel with the hubs. The primary reason is simplicity, as it takes out a class of confounding factors that might cause test failures in staging. In particular, nginx-ingress is the way any traffic (including our hub health tests) get into the cluster, and that changing at the same time as the staging hub deployments is an added operational complexity we can avoid.

> I don't think it's perfect, but I'm at least comfortable it will work 😂 

This is also simple enough for me to understand, so I'd love for us to just do what your proposal is to move things forward, and see where we can go from there?

**Comment by yuvipanda** [comment]
I think https://github.com/2i2c-org/infrastructure/issues/1131#issuecomment-1074232509 (+ support changes) will also perhaps put us in a steady state for a while, and we can focus on working holistically on health checks and alerting separately.

**Comment by yuvipanda** [comment]
> ### Disadvantages
> 
> A failure in job 3 (staging hubs) would prevent any jobs in job 4 (prod hubs) from executing

I also think this is a major *advantage*, rather than a disadvantage! It's the prod hubs doing their job.

**Comment by yuvipanda** [comment]
> For this I meant on a _per cluster_ basis. Do you not want to support the prod hubs to the 2i2c cluster because only pangeo-hubs staging failed, for example?

Ah, sorry I missed that. Ok that clearly is a disadvantage over the current setup. But I think overall, that's an improvement from the current setup, so I'd be +1 on moving ahead as you'd outlined in https://github.com/2i2c-org/infrastructure/issues/1131#issuecomment-1074232509 (- parallel support deployment). We can work on that as a follow up, and I think this will also force us to pay better attention to our health checks.

**Comment by sgibson91** [comment]
Hah, sorry I don't know of any, but it sounded like you did @consideRatio .

Ok, so new plan

### Job 1: Plan jobs

Based on changed files, decide which clusters need something redeploying in, and what exactly needs redeploying. Output some environment variables that describe these jobs when read by the `strategy.matrix` field.

### Job 2: Upgrade support (if needed) and staging hub

```
needs: [job_1]
strategy.matrix = [{cluster_name: some_cluster, upgrade_support: true/false}, ...]
```

If needed, upgrade the support chart. Upgrade the staging hub deployed on a given cluster in parallel. We only upgrade the staging hubs on clusters that will also have production hubs upgraded, this will be decided in Job 1. (Special case for 2i2c which has staging and dask-staging.)

### Job 3: Deploy prod hubs in parallel

```
needs: [job_1, job_2]
strategy.matrix = [{cluster_name: some_cluster, hub_name: hub_1}, {cluster_name: some_cluster, hub_name: hub_2}, {cluster_name: different_cluster, hub_name: hub_3}, ...]
```

Upgrade the production hubs that require an upgrade for each cluster in parallel.

**Comment by yuvipanda** [comment]
> If needed, upgrade the support chart. Upgrade the staging hub deployed on a given cluster in parallel.

To clarify, this would mean the support chart would be upgraded first (if needed), and then *all the staging hubs* will be upgraded in parallel. Not staging hubs and support would be done in parallel. Is that correct?

**Comment by consideRatio** [comment]
@yuvipanda @sgibson91 got time for a call?

**Comment by sgibson91** [comment]
> > If needed, upgrade the support chart. Upgrade the staging hub deployed on a given cluster in parallel.
> 
> To clarify, this would mean the support chart would be upgraded first (if needed), and then _all the staging hubs_ will be upgraded in parallel. Not staging hubs and support would be done in parallel. Is that correct?

We're parallelising over clusters here, so _for each cluster, first upgrade support, then upgrade staging_

> @yuvipanda @sgibson91 got time for a call?

Just sitting down with dinner!

**Comment by yuvipanda** [comment]
@sgibson91 please don't feel the need to work in the evening! I'll just catch up with Erik so I've a better understanding of where things are, and we can all chat tomorrow at a more appropriate time if needed!

And thanks for the clarification re: staging! With that, this plan sounds good to me.

**Comment by consideRatio** [comment]
Wiee I think we could consider this issue as resolved about now!

1. We have a clear strategy to reason about: https://github.com/2i2c-org/infrastructure/issues/1131#issuecomment-1074326220
2. We have verified this can make it all the way to a appreciated end state
   @yuvipanda and I spoke about "job 3a" strategy in https://github.com/2i2c-org/infrastructure/issues/1131#issuecomment-1072573736 and agree its doable and reasonable without adding too much complexity, and further, that this can be done in parallel to any logic to be implemented in the deployer script.
3. We seem to be in agreement that should go for a single workflow kind of solution

I can add further detail to job 3a trick etc, but I figure I'd appreciate just postponing all further work in review/implementation to a planned meet with Sarah in London ~11th April or about that time.

**Comment by yuvipanda** [comment]
Thanks a lot for taking the time to bring me up to speed, @consideRatio!

I'm happy for this to be put on hold until you two meet! I'm also happy to take some load off @consideRatio and do review / discussion work with @sgibson91 if she wants to continue working on it. My goal is to basically offer my own time where possible to move things forward and alleviate load on other team members, if that will help. But this isn't a currently burning fire, so very happy for everyone to take a break. 

To add on to @consideRatio's point, I think we also agreed that https://github.com/2i2c-org/infrastructure/issues/1131#issuecomment-1074326220 is a pretty good next step plan, and there's nothing there that'll prevent 3a of https://github.com/2i2c-org/infrastructure/issues/1131#issuecomment-1072573736 from being implemented after the fact.

I've also come to realize that the flaky health checks are a really big reason why our CD is currently not as trustworthy as we'd like, since we don't actually believe any of the failures it has to be real. I'll try to think about how we can tackle that, but that's outside the scope of this particular issue.

Thank you for all the work you've both put into this, @consideRatio and @sgibson91!
